internal/transport/http2_server.go

```diff
@@ -375,16 +375,39 @@ func (t *http2Server) operateHeaders(frame *http2.MetaHeadersFrame, handle func(
 		}
 	}
 
-	if !isGRPC || headerError {
+	if frame.StreamEnded() && !isGRPC || headerError {
 		t.controlBuf.put(&cleanupStream{
 			streamID: streamID,
 			rst:      true,
-			rstCode:  http2.ErrCodeProtocol,
+			rstCode:  http2.ErrCodeNo,
 			onWrite:  func() {},
 		})
 		return false
 	}
 
+	if !isGRPC {
+		t.controlBuf.put(&headerFrame{
+			streamID: streamID,
+			hf: []hpack.HeaderField{
+				{Name: ":status", Value: "415"},
+			},
+			endStream: true,
+		})
+		return false
+	}
+
+	if headerError {
+		t.controlBuf.put(&headerFrame{
```

Comment:
> In the headerError case, the client is apparently speaking gRPC, so we could return a gRPC status (INTERNAL or UNKNOWN).

Reply:
If we are writing raw headers here, there is more that needs to be included to be a valid gRPC response, right?  E.g. `:status 200` and `grpc-message`.  Can we share some logic with `WriteStatus` to make sure we include all the required fields?

Reply:
I made a call to `WriteStatus` instead of returning raw headers.